Libraries/LibELF/ELFDynamicObject.cpp

```diff
@@ -0,0 +1,602 @@
+#include <AK/StringBuilder.h>
+#include <LibELF/ELFDynamicObject.h>
+
+#include <mman.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define DYNAMIC_LOAD_DEBUG
+//#define DYNAMIC_LOAD_VERBOSE
+
+#ifdef DYNAMIC_LOAD_VERBOSE
+#    define VERBOSE(fmt, ...) dbgprintf(fmt, ##__VA_ARGS__)
+#else
+#    define VERBOSE(fmt, ...) do { } while (0)
+#endif
+
+static bool s_always_bind_now = true;
+
+static const char* name_for_dtag(Elf32_Sword tag);
+
+// SYSV ELF hash algorithm
+// Note that the GNU HASH algorithm has less collisions
+static uint32_t calculate_elf_hash(const char* name)
+{
+    uint32_t hash = 0;
+    uint32_t top_nibble_of_hash = 0;
+
+    while (*name != '\0') {
+        hash = hash << 4;
+        hash += *name;
+        name++;
+
+        top_nibble_of_hash = hash & 0xF0000000U;
+        if (top_nibble_of_hash != 0)
+            hash ^= top_nibble_of_hash >> 24;
+        hash &= ~top_nibble_of_hash;
+    }
+
+    return hash;
+}
+
+RefPtr<ELFDynamicObject> ELFDynamicObject::make(const char* filename, int fd, size_t size)
+{
+    return { new ELFDynamicObject(filename, fd, size) };
+}
+
+ELFDynamicObject::ELFDynamicObject(const char* filename, int fd, size_t size)
+    : m_filename(filename)
+    , m_file_size(size)
+    , m_image_fd(fd)
+{
+    String file_mmap_name = String::format("ELF_DYN: %s", m_filename.characters());
+
+    m_file_mapping = mmap_with_name(nullptr, size, PROT_READ, MAP_PRIVATE, m_image_fd, 0, file_mmap_name.characters());
+    if (MAP_FAILED == m_file_mapping) {
+        m_valid = false;
+        return;
+    }
+
+    m_image = AK::make<ELFImage>((u8*)m_file_mapping);
+
+    m_valid = m_image->is_valid() && m_image->parse() && m_image->is_dynamic();
+
+    if (!m_valid) {
+        return;
+    }
+
+    const ELFImage::DynamicSection probably_dynamic_section = m_image->dynamic_section();
+    if (StringView(".dynamic") != probably_dynamic_section.name() || probably_dynamic_section.type() != SHT_DYNAMIC) {
+        m_valid = false;
+        return;
+    }
+}
+
+ELFDynamicObject::~ELFDynamicObject()
+{
+    if (MAP_FAILED != m_file_mapping)
+        munmap(m_file_mapping, m_file_size);
+}
+
+void ELFDynamicObject::dump()
+{
+    auto dynamic_section = m_image->dynamic_section();
+
+    StringBuilder builder;
+    builder.appendf("\nd_tag      tag_name         value\n");
```

Comment:
This can just use `append()` since there's no format string action.

Reply:
true, no need to try and parse 0 vargs :)